04-rag-demo/README.md

```diff
@@ -0,0 +1,135 @@
+
+## Setup
+
+You can use this 1-line command to start the RisingWave server.
+
+```bash
+curl -L https://risingwave.com/sh | sh
+```
+
+After the server started, create the necessary UDFs and tables.
+
+The following SQL creates a UDF `text_embedding` that uses OpenAI's API to embed a text into a vector. The function relies on the `fetch` API, which is introduced in the latest version of RisingWave.
+
+```sql
+create function text_embedding(t varchar) returns real[] language javascript as $$
+    export async function text_embedding(t) {
+        for (let retries = 0; retries < 3; retries++) {
+            try {
+                const response = await fetch("https://api.openai.com/v1/embeddings", {
+                    method: 'POST',
+                    headers: {
+                        'Authorization': `Bearer <your-openai-api-key>`,
+                        'Content-Type': 'application/json'
+                    },
+                    body: JSON.stringify({
+                        input: t,
+                        model: "text-embedding-3-small"
+                    })
+                });
+                if (!response.ok) {
+                    const error = await response.json();
+                    throw new Error(`OpenAI API error: ${error.error?.message || 'Unknown error'}`);
+                }
+                const data = await response.json();
+                return data.data[0].embedding;
+            } catch (error) {
+            }
+        }
+        return null;
+    }
+$$ WITH (
+    async = true,
+    batch = false
+);
+```
+
+Now, create the data pipeline to ingest the documents into the database. In RisingWave, a data pipeline is simply written as a materialized view defined in SQL.
+
+```sql
+CREATE TABLE documents (
+    file_name VARCHAR,
+    content TEXT,
+    PRIMARY KEY (file_name)
+);
+
+CREATE MATERIALIZED VIEW document_embeddings AS
+WITH t AS (
+    SELECT
+        *, text_embedding(content) AS embedding
+    FROM documents
+)
+SELECT
+    file_name,
+    content,
+    embedding
+FROM t
+WHERE embedding IS NOT NULL;
+```
+
+## Load data
+
+For this demo, we use the documents from the RisingWave docs. 
+
+```bash
+git clone https://github.com/risingwavelabs/risingwave-docs.git
+cd risingwave-docs
+
+# Load all the documents (`.mdx` files) into the `documents` table
+find . -name "*.mdx" | while read -r file; do
+    content=$(cat "$file" | sed "s/'/''/g")
+    sql_statement="INSERT INTO documents (file_name, content) VALUES ('$file', '$content') ON CONFLICT (file_name) DO UPDATE SET content = EXCLUDED.content;"
+    echo "$sql_statement" | psql -h 127.0.0.1 -p 4566 -d dev -U root
+    echo "Processed: $file"
+done
+```
+
+## Query
+
+To compare the similarity between the question and the documents, we need to introduce the `cosine_similarity` UDF.
+
+```sql
+create function cosine_similarity(v1 real[], v2 real[]) returns real language rust as $$
```

Comment:
Shall we use capital form of keywords in this and above `create function` blocks? Just for consistency.